File: src/engine/core/tree.ts
import { EngineEventListener, GameObject } from ".";

export class Tree extends EngineEventListener {
    readonly root: GameObject;

    private readonly id_gameObject: Map<number, GameObject>;

    constructor() {
        super();

        this.root = new GameObject();
        this.id_gameObject = new Map();
    }

    getGameObjectByID(ID: number) {
        return this.id_gameObject.get(ID);
    }

    addChild(parent: GameObject, child: GameObject) {
        // assert: parent doesn't already have this child

        if(child.parent)
            this.removeChild(child.parent, child);

        this.id_gameObject.set(child.ID, child);
        parent.children.push(child);
        child.parent = parent;
        child.visible = parent.visible;
    }

    removeChild(parent: GameObject, child: GameObject) {
        // assert: index is always >= 0
        const index = parent.children.indexOf(child);

        this.id_gameObject.delete(child.ID);
        parent.children.splice(index, 1);

        child.parent = null;
    }

    addGameObject(gameObject: GameObject) {
        this.id_gameObject.set(gameObject.ID, gameObject);
        this.addChild(this.root, gameObject);
    }

    static createGameObject() {
        return new GameObject();
    }

    spawnGameObject() {
        const gameObject = Tree.createGameObject();

        this.addGameObject(gameObject);

        return gameObject;
    }

    traverse(callback: (gameObject: GameObject) => void) {
        Tree.traverseChildren(this.root, callback);
    }

    updateTransforms() {
        this.traverse(gameObject => gameObject.transform.updateAbsoluteValues());
    }

    static traverseChildren(gameObject: GameObject, callback: (gameObject: GameObject) => void) {
        const q = [gameObject];

        while(q.length) {
            const n = q[0];

            callback(n);

            q.shift();
            q.push(...n.children);
        }
    }
}

File: src/server/scene/tree.cpp
#include "tree.hpp"

#include <queue>

Tree::Tree() {
    this->root = make_shared<GameObject>();
}

shared_ptr<GameObject> Tree::getGameObjectByID(const int ID) {
    return id_gameObject[ID];
}

void Tree::addChild(shared_ptr<GameObject>& parent, shared_ptr<GameObject>& child) {
    if(child->parent)
        removeChild(child->parent, child);

    id_gameObject[child->ID] = child;
    parent->children.insert(child);
    child->parent = parent;
}

void Tree::removeChild(shared_ptr<GameObject>& parent, shared_ptr<GameObject> &child) {
    id_gameObject.erase(child->ID);
    parent->children.erase(child);

    child->parent = nullptr;
}

void Tree::addGameObject(shared_ptr<GameObject>& gameObject) {
    id_gameObject[gameObject->ID] = gameObject;
    addChild(root, gameObject);
}

shared_ptr<GameObject> Tree::createGameObject() {
    return make_shared<GameObject>(GameObject());
}

shared_ptr<GameObject> Tree::spawnGameObject() {
    auto gameObject = createGameObject();

    addGameObject(gameObject);

    return gameObject;
}

void Tree::traverse(const function<void(shared_ptr<GameObject>& gameObject)>& callback) {
    traverseChildren(root, callback);
}

void Tree::traverseChildren(
    const shared_ptr<GameObject>& gameObject,
    const function<void(shared_ptr<GameObject>&)>& callback
) {
    queue<shared_ptr<GameObject>> q;
    q.push(gameObject);

    while(!q.empty()) {
        shared_ptr<GameObject> current = q.front();
        q.pop();

        callback(current);

        for(const auto& child: current->children)
            q.push(child);
    }
}

void Tree::traverseDirectChildren(
    const shared_ptr<GameObject>& gameObject,
    const function<void(const shared_ptr<GameObject>&)>& callback
) {
    for(auto& child: gameObject->children) {
        callback(child);
    }
}

void updateTransform(shared_ptr<GameObject>& gameObject) {
    if(gameObject->parent == nullptr)
        return;

    gameObject->transform->updateAbsoluteValues(gameObject->parent->transform);
}

void Tree::updateTransforms() {
    traverse(updateTransform);
}

void Tree::updateTransforms(shared_ptr<GameObject>& gameObject) {
    traverseChildren(gameObject, updateTransform);
}
